content/blog/entries/crawling-500-million/contents.lr

```diff
@@ -0,0 +1,485 @@
+
+We solved (3) by setting up a [caching thumbnail proxy](https://github.com/willnorris/imageproxy) between images in the search results and their 3rd party origin, as well as some last-minute liveness checks to make sure that the image hasn't 404'd.
+
+(1) and (2), however, are not possible to solve without actually downloading the image and performing some analysis on the contents of the file. For us to reproduce the features that users take for granted in image search, we're going to need a fairly powerful crawling system.
+
+On the scale of several thousand images, it would be easy to cobble together a few scripts to spit out this information, but with half a billion images, there are a lot of hurdles to overcome.
+
+- We want to crawl [politely](https://en.wikipedia.org/wiki/Web_crawler#Politeness_policy); however, the concentration and quantity of images means that we have to hit some sources with a high crawl rate in order to have any hope of finishing the crawl in a reasonable period of time. Our data sources range from non-profit museums with a single staff IT person to tech companies with their own data centers and thousands of employees; the crawl rate has to be tailored to download quickly from the big players but not overwhelm small sources. At the same time, we need to be sure that we are not overestimating any source's capacity and watch for signs that our crawler is straining the server.
+- We need to keep the time to process each image as low as possible to make it feasible to finish the crawling and analysis task in a reasonable period of time. This means that the crawling and analysis tasks need to be distributed to multiple machines in parallel.
+- A lot of metadata will be produced by this crawler. The step of integrating it with our internal systems needs to not block resizing tasks. That suggests that a message bus will be necessary to buffer messages before they are written into our data layer, where writes can be expensive.
+- We want to have a basic idea of how the crawl is progressing in the form of summaries of error counts, status codes, and crawl rates for each source.
+
+To summarize, the challenge here isn't so much making a really fast crawler as much as it is tailoring the crawl speed to each source. At a minimum, we'll need to deal with concurrency and parallelism, provisioning and managing the life cycle of crawler infrastructure, pipelines for capturing output data, a way to monitor the progress of the crawl, a suite of tests to make sure the system behaves as expected, and a reliable way to enforce a "politeness" policy. That's not a trivial project, particularly for our tiny three person tech team (of which only one person is available to do all of the crawling work). Can't we just use an off-the-shelf open source crawler?
+
+## What about existing open source crawlers?
```

Comment:
It may be worth adding information about what open source crawlers you considered.

Reply:
I held back on naming any specific frameworks because I didn't want to accuse anyone of being in "a poorer state of maintenance than one would think at first glance"

Reply:
Perhaps find a more diplomatic way to put it? I think it would be useful for readers to know what limitations of existing crawlers are, even if it's just in our opinion.

Reply:
I rephrased that sentence and listed the three crawling frameworks I tried out